!!! FEATURE: `Neos.Neos:Site` NodeType

mhsdesign opened the pull request:
<!-- 
    Thanks for your contribution, we appreciate it!

    The first section should explain briefly what is changed. 
    Some examples are always nice to showcase the use. 
    The content will be used in the change-logs and addresses 
    developers working with Neos.

    If there are issues regarding the topic of your PR link 
    them here as `related:` or `resolved:`
-->

resolves: neos/neos-development-collection#4531

**Upgrade instructions**

Your homepage must extend `Neos.Neos:Site`

## For more upgrade information see https://github.com/neos/neos-development-collection/issues/5400

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->

**Review instructions**

Todo actually use the `Neos.Neos:Site` in the php code when querying.

<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mhsdesign committed:
FEATURE: `Neos.Neos:Site` NodeType

mhsdesign committed:
TASK: Remove unused exceptions

mhsdesign committed:
TASK: Assert in `site:create` that homepage node must be of type `Neos.Neos:Site`

mhsdesign committed:
TASK: Correctly stringify NodeTypeName

mhsdesign committed:
TASK: Adjust legacy node migrations to enforce that site node must be of type `Neos.Neos:Site`.

mhsdesign committed:
TASK: Guard `NodeDataToEventsProcessor::setSitesNodeType`

kitsunet commented:
Makes sense I guess. I am sure though there will be many sites out there that do not have an explicit homepage nodetype right now and need structure adjustments for this. Do we / can we prepare one for this? Or a command ala `turnsitenodeintonodetype <your.node:type>` ? Could probably also happen during converting from old to new.

mhsdesign commented:
> Neos.ContentRepository.LegacyNodeMigration/Tests/Behavior/behat.yml.dist
>    AfterScenario # FeatureContext::failIfLastMigrationHasErrors()
>      RuntimeException: The last migration run led to an error: The site node "site-node-id" (type: "Some.Package:SomeNodeType") must be of type "Neos.Neos:Site". in Bootstrap/FeatureContext.php:90

mhsdesign commented:
About an upgrade guide for people without dedicated page node i have created this task: https://github.com/neos/neos-development-collection/issues/5400

dlubitz commented on Neos.ContentRepository.LegacyNodeMigration/Classes/NodeDataToEventsProcessor.php:
```diff
@@ -87,7 +86,7 @@ public function __construct(
         private readonly iterable $nodeDataRows,
     ) {
-        $this->sitesNodeTypeName = NodeTypeName::fromString('Neos.Neos:Sites');
+        $this->sitesNodeTypeName = NodeTypeName::fromString(NodeTypeNameFactory::NAME_SITES);
```
```suggestion
        $this->sitesNodeTypeName = NodeTypeName::forSites();
```

mhsdesign replied:
yes for this case i very much agree to use this ^^ i was a bit blinded.

dlubitz reviewed (commented):
I don't like the mix up of public constant and static method usage. Would be more clear if there were just one way of getting the nodeTypeName values. Even if it's more complex to get it in cases where you simply need the string value. But this is the price of type safetiness, isn't it?

mhsdesign committed:
TASK: Use `NodeTypeName::forSites` correctly

Co-authored-by: Denny Lubitz <lubitz@vivomedia.de>

mhsdesign commented:
> I don't like the mix up of public constant and static method usage.

i thought so too at first, but after seeing `NodeTypeName::forSites()->value` i was thankful for `NodeTypeNameFactory::NAME_SITES` ^^

But either way lets discuss this in a dedicated issue to not block this pr :D

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into feature/NeosNeosSiteNodeType

mhsdesign committed:
TASK: Adjust Legacy migration tests to `Neos.Neos:Site`

mhsdesign committed:
TASK: Legacy migration test scenario: Homepage node is not of type "Neos.Neos:Site"

mhsdesign committed:
TASK: Adjust `SiteNodeUtility` to use 'Neos.Neos:Sites'

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into feature/NeosNeosSiteNodeType

mhsdesign commented:
I resolved merge conflicts and tested this locally with the neos demo (adjustment https://github.com/neos/Neos.Demo/pull/182) and the neos ui (adjustment https://github.com/neos/neos-ui/pull/3629)

the neos ui e2e test succeed locally.

mhsdesign committed:
TASK: Inline `SiteNodeUtility::findSiteNode`

mhsdesign committed:
TASK: Use named arguments for `FindClosestNodeFilter`

mhsdesign committed:
TASK: Utilize `Neos.Neos:Site` at various places

mhsdesign committed:
FEATURE: Add `SiteNodeUtility::findSiteNodeBySite`

The rather hacky method `findCurrentSiteNode` was refactored and its usage adjusted.

mhsdesign committed:
TASK: Fix ci

nezaniel commented:
Looks nice, I'll add the document uri path projection adjustments in a later pr

bwaidelich commented on Neos.Neos/NodeTypes/Mixin/Site.yaml:
```diff
@@ -0,0 +1,10 @@
+'Neos.Neos:Site':
+  abstract: true
+  superTypes:
+    'Neos.Neos:Document': true
```
We briefly discussed this in our weekly:
Actually the `uriPathSegment` property should not be part of the homepage.
It would be rather nasty to remove the property from this type even though it is declared in the super type.. But I would say that this is the most pragmatic solution in this case. _If_ the following works:

```suggestion
    'Neos.Neos:Document': true
  properties:
    'uriPathSegment': ~
```

(untested!)

Otherwise we should at least disable the UI editor for this property

kitsunet replied:
I do not like it, but I have to admit I don't really see a point were this goes super bad, so lets do it :)

mhsdesign replied:
I created a separate issue for this: https://github.com/neos/neos-development-collection/issues/4621

as it seems we might not like the unsetting https://github.com/neos/neos-development-collection/pull/4618

bwaidelich replied:
I agree, thanks!
But maybe we can at least hide it from the inspector by default?
